### ISSUE TITLE
bug in `fftfeq_to_dft_coordinates`

### ISSUE THREAD

**Issue by alisterburt** [teamtomo/torch-fourier-slice#28]
the fftfreq -> dft sample coordinate conversion is slightly off, it assumes fftfreq over depth and height dimensions covers `[-0.5, 0.5]` when it actually covers `[-0.5, 0.5 - step]` where step is the step between dft sample frequencies along the dimension. 

The error is small and gets smaller as number of samples gets larger

cc @McHaillet @jdickerson95 @mgiammar @rsanchezgarc as known consumers of this library

**Comment by mgiammar** [comment]
Oh man good catch how did you figure this out? Four our (me and @jdickerson95) sampling is usually on grids of 256-512 pixels, so I'd imagine it only being a _slight_ issue. Can check how much this affects some of our projections.

**Comment by alisterburt** [comment]
@mgiammar I [rewrote this for the GPU in Mojo](https://github.com/alisterburt/mojo-cryoem-projection) over the weekend and noticed the rescaling logic here assumed an interval width of 1 covered by the the full dimensions in fftfreq sample space

```ipython
In [1]: import numpy as np

In [2]: np.fft.fftshift(np.fft.fftfreq(10))
Out[2]: array([-0.5, -0.4, -0.3, -0.2, -0.1,  0. ,  0.1,  0.2,  0.3,  0.4])

In [3]: np.fft.fftshift(np.fft.fftfreq(256))
Out[3]: 
array([-0.5       , -0.49609375, -0.4921875 , -0.48828125, -0.484375  ,
       -0.48046875, -0.4765625 , -0.47265625, -0.46875   , -0.46484375,
        ...
        0.4765625 ,  0.48046875,  0.484375  ,  0.48828125,  0.4921875 ,
        0.49609375])
```

the difference is small but not nothing!

**Comment by alisterburt** [comment]
this line is the one with the error
https://github.com/teamtomo/torch-fourier-slice/blob/20a995e40e248f3cb1275909263e9b8361049399/src/torch_fourier_slice/_dft_utils.py#L80

**Comment by mgiammar** [comment]
I briefly took a look at this and tried a very hacky bugfix which changes the coordinates _before_ they're rotated by adding the following lines to the extract central slices function

https://github.com/teamtomo/torch-fourier-slice/blob/20a995e40e248f3cb1275909263e9b8361049399/src/torch_fourier_slice/slice_extraction/_extract_central_slices_rfft_3d.py#L44-L45

new added lines:

```python
    print(valid_coords.shape)
    print(f"valid_coords: {valid_coords[..., 0]}")

    # BUGFIX Adjust spacing along y-axis to be -0.5 to 0.5 instead of -0.5 to (0.5-inc)
    last_dim_shape = int(rfft_shape[-1])
    full_dim_shape = int(rfft_shape[-2])
    adjusted_freq = torch.linspace(-0.5, 0.5, full_dim_shape)

    for i in range(full_dim_shape):
        start_index = i * last_dim_shape
        end_index = start_index + last_dim_shape

        # print(valid_coords[start_index:end_index, 1, 0].size(0))

        assert torch.unique(valid_coords[start_index:end_index, 1, 0]).size(0) == 1

        valid_coords[start_index:end_index, 1, 0] = adjusted_freq[i]
```

However, results for a 60S template make it seem like what is currently implemented is somehow not affected by the above issue. Plots are below:

![Image](https://github.com/user-attachments/assets/2dc8a59c-c716-4cc8-9738-7addefd240e3)

![Image](https://github.com/user-attachments/assets/d71e2561-6312-4a04-9b93-c1713b54bd7b)

**Comment by mgiammar** [comment]
The "fix" I listed above is very possibly wrong, but it could be the case that the subtracted increment(s) are somehow being handled by the coordinate normalization to `[-1, 1]` before being passed to [torch.nn.functional.grid_sample](https://docs.pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html#torch-nn-functional-grid-sample).

**Comment by alisterburt** [comment]
@mgiammar thanks for taking a look! I think the correct logic is to calculate the dim length as `1 - delta_fftfreq`

```python
In [7]: N = 10
In [8]: freqs = np.fft.fftfreq(N)
In [9]: delta_fftfreq = 1 / N
In [10]: np.max(freqs) - np.min(freqs)
Out[10]: np.float64(0.9)
In [11]: 1 - delta_fftfreq
Out[11]: 0.9
```

The fftfreq -> sample position in dft calculation then looks like
```
yrot = ((yrot + 0.5) / y_dim_length) * (h - 1)
zrot = ((zrot + 0.5) / z_dim_length) * (d - 1)
```

`(yrot + 0.5) / y_dim_length)` correctly takes the values to [0, 1] then `* (h - 1)` scales to `[0, h-1]`

What we do rescales first, assuming a dim length of 1

**Comment by alisterburt** [comment]
@rsanchezgarc hm no I'm not saying that - the original fft frequency grid construction is correct, the way this maps to positions in the DFT which are sampled is *slightly* incorrect

**Comment by McHaillet** [comment]
Nice that you figured this out! I have used the functions with smaller box sizes, so it should have an impact :)
